src/proc/constants.rs

```diff
@@ -0,0 +1,564 @@
+            Expression::Call { .. } => Err(ConstantSolvingError::Call),
+            Expression::FunctionArgument(_) => Err(ConstantSolvingError::FunctionArg),
+            Expression::GlobalVariable(_) => Err(ConstantSolvingError::GlobalVariable),
+            Expression::ImageSample { .. } => Err(ConstantSolvingError::ImageExpression),
+            Expression::ImageLoad { .. } => Err(ConstantSolvingError::ImageExpression),
+        }
+    }
+
+    fn access(
+        &mut self,
+        base: Handle<Expression>,
+        index: usize,
+    ) -> Result<Handle<Constant>, ConstantSolvingError> {
+        let base = self.solve(base)?;
+
+        match self.constants[base].inner {
+            crate::ConstantInner::Scalar { .. } => Err(ConstantSolvingError::InvalidAccessBase),
+            crate::ConstantInner::Composite { ty, ref components } => match self.types[ty].inner {
+                crate::TypeInner::Vector { size, .. } => {
+                    if size as usize <= index {
+                        Err(ConstantSolvingError::InvalidAccessIndex)
+                    } else {
+                        Ok(components[index])
+                    }
+                }
+                crate::TypeInner::Matrix {
+                    rows,
+                    columns,
+                    width,
+                } => {
+                    if columns as usize <= index {
+                        return Err(ConstantSolvingError::InvalidAccessIndex);
+                    }
+
+                    let components =
+                        components[(index * rows as usize)..((index + 1) * rows as usize)].to_vec();
+
+                    let ty = self.types.fetch_or_append(Type {
+                        name: None,
+                        inner: TypeInner::Vector {
+                            size: rows,
+                            kind: ScalarKind::Float,
+                            width,
+                        },
+                    });
+
+                    Ok(self.constants.fetch_or_append(Constant {
+                        name: None,
+                        specialization: None,
+                        inner: ConstantInner::Composite { ty, components },
+                    }))
+                }
+                crate::TypeInner::Array { size, .. } => match size {
+                    ArraySize::Constant(constant) => {
+                        let size = self.constant_index(constant)?;
+
+                        if size <= index {
+                            Err(ConstantSolvingError::InvalidAccessIndex)
+                        } else {
+                            Ok(components[index])
+                        }
+                    }
+                    ArraySize::Dynamic => Err(ConstantSolvingError::ArrayLengthDynamic),
+                },
+                crate::TypeInner::Struct { ref members, .. } => {
+                    if members.len() <= index {
+                        Err(ConstantSolvingError::InvalidAccessIndex)
```

Comment:
nit: you can do a fancy `components.get(index).map_or(ConstantSolvingError::InvalidAccessIndex).cloned()` here to avoid bounds-checking twice